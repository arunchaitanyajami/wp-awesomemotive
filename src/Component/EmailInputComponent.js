import { useState, useEffect } from '@wordpress/element';
import { __experimentalInputControl as InputControl } from '@wordpress/components';
import { __, sprintf } from '@wordpress/i18n';
import * as defaultOptions from '../deafultOptions';
import { Icon, plus, trash } from '@wordpress/icons';

/**
 * Email input component.
 *
 * @param props
 * @returns {JSX.Element}
 * @constructor
 */
export default ( { value, keyId, saveEmail, removeEmail } ) => {
	const [ email, setEmail ] = useState( value );
	const [ invalidEmail, setInvalidEmail ] = useState( false );

	const preCheckEmail = ( value ) => {
		let isEmail = value.match(
			/^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
		);

		setEmail( value );
		setInvalidEmail( ! isEmail );
		if ( isEmail ) {
			saveEmail( keyId - 1, value );
		}
	};

	const deleteEmail = () => {
		removeEmail( keyId - 1 );
	};

	return (
		<div
			className={ 'awesomemotive-settings-form-input-repeater' }
			id={ 'awesomemotive-settings-form-input-repeater' + keyId }
		>
			{ invalidEmail && email && (
				<div className={ 'awesomemotive-settings-form-input-error' }>
					{ __(
						'Invalid email address',
						defaultOptions.text_domain
					) }{ ' ' }
				</div>
			) }
			<div className={ 'awesomemotive-settings-form-input-control' }>
				<InputControl
					label={ __(
						sprintf( 'Email %d', keyId ),
						defaultOptions.text_domain
					) }
					value={ email }
					type={ 'email' }
					name={ 'email' }
					id={ sprintf( 'email_%d', keyId ) }
					required
					placeholder={ 'Enter email address..' }
					onChange={ ( nextValue ) =>
						preCheckEmail( nextValue ?? '' )
					}
				/>
			</div>
			<div className={ 'awesomemotive-settings-form-input-trash-icon' }>
				<Icon icon={ trash } onClick={ () => deleteEmail() } />
			</div>
		</div>
	);
};
