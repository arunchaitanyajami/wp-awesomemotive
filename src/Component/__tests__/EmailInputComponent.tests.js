import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import EmailInputComponent from '../EmailInputComponent';

let props = {
	value: 'test@test.com',
	keyId: 1,
	keyIndex: 1,
	saveEmail: jest.fn(),
	removeEmail: jest.fn(),
};

describe( 'Email Input Component', () => {
	let useEffect;
	let useState;
	beforeEach( () => {
		useEffect = jest.spyOn( React, 'useEffect' );
		useState = jest.spyOn( React, 'useState' );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	const mockUseEffect = () => {
		useEffect.mockImplementationOnce( ( f ) => f() );
	};

	it( 'Should Render the component', async () => {
		useState
			.mockReturnValueOnce( [ props.value, jest.fn() ] )
			.mockReturnValueOnce( [ false, jest.fn() ] )
			.mockReturnValueOnce( [ props.keyId, jest.fn() ] );

		const { container } = render( <EmailInputComponent { ...props } /> );

		expect( container.querySelector( '#email_1' ).value ).toBe(
			props.value
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'Should Render the component with error', async () => {
		useState
			.mockReturnValueOnce( [ props.value, jest.fn() ] )
			.mockReturnValueOnce( [ true, jest.fn() ] )
			.mockReturnValueOnce( [ props.keyId, jest.fn() ] );

		const { container } = render( <EmailInputComponent { ...props } /> );

		expect(
			container.querySelector(
				'.awesomemotive-settings-form-input-error'
			)
		).toBeInTheDocument();
		expect(
			screen.getByText( 'Invalid email address' )
		).toBeInTheDocument();
		expect( container ).toMatchSnapshot();
	} );
} );
