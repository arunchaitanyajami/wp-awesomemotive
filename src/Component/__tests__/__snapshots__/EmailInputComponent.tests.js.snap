// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email Input Component Should Render the component 1`] = `
<div>
  <div
    class="awesomemotive-settings-form-input-repeater"
    id="awesomemotive-settings-form-input-repeater1"
  >
    <div
      class="awesomemotive-settings-form-input-control"
    >
      <div
        class="components-flex components-input-control em5sgkm7 css-sh7xzs-View-Flex-sx-Base-sx-Items-ItemsColumn-Root-rootFocusedStyles-rootLabelPositionStyles em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Flex"
      >
        <div
          class="components-flex-item em5sgkm3 css-1fmchc6-View-Item-sx-Base-LabelWrapper em57xhy0"
          data-wp-c16t="true"
          data-wp-component="FlexItem"
        >
          <label
            class="components-truncate components-text components-input-control__label em5sgkm4 css-1790joj-View-Text-sx-Base-BaseLabel-labelMargin em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
            for="email_1"
          >
            Email 1
          </label>
        </div>
        <div
          class="components-input-control__container css-1t6cb2b-Container-containerDisabledStyles-containerWidthStyles em5sgkm6"
        >
          <input
            class="components-input-control__input css-gzm6eu-Input-dragStyles-fontSizeStyles-sizeStyles em5sgkm5"
            id="email_1"
            name="email"
            placeholder="Enter email address.."
            required=""
            type="email"
            value="test@test.com"
          />
          <div
            aria-hidden="true"
            class="components-input-control__backdrop css-6ems4a-BackdropUI-backdropFocusedStyles em5sgkm2"
          />
        </div>
      </div>
    </div>
    <div
      class="awesomemotive-settings-form-input-trash-icon"
    >
      <svg
        aria-hidden="true"
        focusable="false"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 5h-5.7c0-1.3-1-2.3-2.3-2.3S9.7 3.7 9.7 5H4v2h1.5v.3l1.7 11.1c.1 1 1 1.7 2 1.7h5.7c1 0 1.8-.7 2-1.7l1.7-11.1V7H20V5zm-3.2 2l-1.7 11.1c0 .1-.1.2-.3.2H9.1c-.1 0-.3-.1-.3-.2L7.2 7h9.6z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Email Input Component Should Render the component with error 1`] = `
<div>
  <div
    class="awesomemotive-settings-form-input-repeater"
    id="awesomemotive-settings-form-input-repeater1"
  >
    <div
      class="awesomemotive-settings-form-input-error"
    >
      Invalid email address
       
    </div>
    <div
      class="awesomemotive-settings-form-input-control"
    >
      <div
        class="components-flex components-input-control em5sgkm7 css-sh7xzs-View-Flex-sx-Base-sx-Items-ItemsColumn-Root-rootFocusedStyles-rootLabelPositionStyles em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Flex"
      >
        <div
          class="components-flex-item em5sgkm3 css-1fmchc6-View-Item-sx-Base-LabelWrapper em57xhy0"
          data-wp-c16t="true"
          data-wp-component="FlexItem"
        >
          <label
            class="components-truncate components-text components-input-control__label em5sgkm4 css-1790joj-View-Text-sx-Base-BaseLabel-labelMargin em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
            for="email_1"
          >
            Email 1
          </label>
        </div>
        <div
          class="components-input-control__container css-1t6cb2b-Container-containerDisabledStyles-containerWidthStyles em5sgkm6"
        >
          <input
            class="components-input-control__input css-gzm6eu-Input-dragStyles-fontSizeStyles-sizeStyles em5sgkm5"
            id="email_1"
            name="email"
            placeholder="Enter email address.."
            required=""
            type="email"
            value="test@test.com"
          />
          <div
            aria-hidden="true"
            class="components-input-control__backdrop css-6ems4a-BackdropUI-backdropFocusedStyles em5sgkm2"
          />
        </div>
      </div>
    </div>
    <div
      class="awesomemotive-settings-form-input-trash-icon"
    >
      <svg
        aria-hidden="true"
        focusable="false"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 5h-5.7c0-1.3-1-2.3-2.3-2.3S9.7 3.7 9.7 5H4v2h1.5v.3l1.7 11.1c.1 1 1 1.7 2 1.7h5.7c1 0 1.8-.7 2-1.7l1.7-11.1V7H20V5zm-3.2 2l-1.7 11.1c0 .1-.1.2-.3.2H9.1c-.1 0-.3-.1-.3-.2L7.2 7h9.6z"
        />
      </svg>
    </div>
  </div>
</div>
`;
